### PR TITLE
[Plan to Close this PR] Mainloop multistaged pipeline and instruction scheduling for NVIDIA Ampere Tensor Cores (F16)

### PR DESCRIPTION
**Pull request by manishucsd**
This PR creates a _coarse-grained_ multistage pipelining and _fine-grained_ instruction scheduling for optimal performance on NVIDIA Ampere Tensor Cores. 

Multi-staging is essential to hide the Global Memory load latency by building longer software pipelines and using the available Shared Memory capacity, especially on NVIDIA A100. Additionally, fine-grained instruction scheduling hides the Shared Memory load latency by prefetching the math operands into registers. 

**Note** that this PR removes `mma.sync.1688.f32.tf32` tests. The `mma.sync.1688.f32.tf32` needs fine-grained schedule as well, the support for it added in the next [PR 12603](https://github.com/openxla/iree/pull/12603). Once this is merged, rebase 12630 and it can pushed in as well.

### CODE REVIEW

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp [hunk 1]
```diff
@@ -244,107 +229,292 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
...
+/// Structure to hold the matmul's mainloop information:
+/// Seperates the mmasync operations into kgroups and collects the Shared Memory
+/// loads for each kgroup. This information is used to pipeline the mainloop and
+/// to generate an optimal schedule interleaving Global Memory loads, Shared
+/// Memory loads, and math operations.
+struct MainLoopInfo {
+  //
+  // Type definitions
+  //
+  using KgroupToWarpMmaMap = std::unordered_map<int, WarpMmaOp>;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: why does it need to be an `unordered_map`? You can just make it a SmallVector

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp [hunk 2]
```diff
@@ -280,7 +280,7 @@ void insertBarriersAroundSharedMemoryCopy(func::FuncOp funcOp) {
       Operation *prevOp = copyOp->getPrevNode();
       if (!prevOp || !hasMarker(prevOp, getCopyToWorkgroupMemoryMarker())) {
         builder.setInsertionPoint(copyOp);
-        builder.create<gpu::BarrierOp>(copyOp->getLoc());
+        // builder.create<gpu::BarrierOp>(copyOp->getLoc());
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: remove

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp [hunk 3]
```diff
@@ -365,7 +535,7 @@ static FailureOr<scf::ForOp> applyPipelining(
                          scf::ForOp forOp,
                          std::vector<std::pair<Operation*, unsigned>>& ops) {
     if (schedule == PipeliningSchedulingStrategy::nvidiaTensorCore) {
-      return getNvidiaTensorCorePipeline(forOp, ops, maxDepth);
+      return getMultiStagedPipelineSchedule(forOp, ops, maxDepth);
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Can we keep Nvidia in the name. This is highly specific to this target.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp [hunk 4]
```diff
@@ -244,107 +229,292 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
...
-    if (stageNMinusOneDeps.count(&op))
-      ops.push_back(std::make_pair(&op, depth - 2));
+    if (info.mmasyncOpDeps[numKgroups - 1].count(&op))
+      ops.push_back(std::make_pair(&op, numStages - 1));
   }
 }
 
-/// Apply pipeline rewrite pattern assuming the operations were already
-/// annotated with stage information.
+// Apply pipeline rewrite pattern assuming the operations were already
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:551]: why did you change the comment to `//`, `///` was correct.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp [hunk 5]
```diff
@@ -244,107 +229,292 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
...
+    unsigned numStages) {
+  // Analyze the main loop and obtain information for coarse-grained pipelining
+  // and fine-grained instruction scheduling.
+  MainLoopInfo info(forOp);
+  int numKgroups = info.getNumberOfKgroups();
+
+  // Assert the requirements for pipelining the mainloop targeting NVIDIA
+  // Tensor Cores using multistaged pipeline.
+  assert(numKgroups > 1 && "Number of kgroups should be 2 ore more");
+  assert(numStages > 2 && "Number of stages should be 3 ore more");
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: What guarantees that? We should early exit instead.

**Comment by qcolombet** [comment]
Could you add a test case that demonstrates what the new code is doing?
Maybe this already exists and in that case, you just forgot to update the check lines :).

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp [hunk 6]
```diff
@@ -244,107 +229,292 @@ static bool setPipeliningMarkers(scf::ForOp forOp, bool pipelineStoreStage) {
...
+
+  // Map: Operation -> kgroupIdx
+  std::unordered_map<Operation*, int> opToKgroupMap;
+
+  // Set to track the dependencies already seen to a backward slice.
+  llvm::SetVector<Operation*> seenDepOps;
+
+  // Set to track the mma operations in forward slice to count kgroups and
+  // populate the warp-level WarpMmaOp kgroupIdx -> WarpMmaOp map.
+  llvm::SetVector<Operation*> seenMmaOps;
```
Comment (ThomasRaoux) [inline compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: I think we should be able to simplify those data structure.

Do we really need the reverse mapping `opToKgroupMap`? The mmasync are already sorted by groups, can't you just loop through.
Also can we merge the operations with their dependency into one SetVector rather than having two?

Also in general you want to prefer `llvm::DenseMap` to `std::unordered_map` but I think hopefully we can remove those maps here anyway.

**Comment by ThomasRaoux** [comment]
> Could you add a test case that demonstrates what the new code is doing? Maybe this already exists and in that case, you just forgot to update the check lines :).

+1, please add/update a test (there is already be one for the special pipelining schedule if that doesn't trigger the different IR generates let's add a new one)

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 4 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:551]: Changing it. 

What is the guideline/rule on `//` vs `///`?

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 5 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: `numStages > 2` we an assert before the call, but `numKgroups` info we only have it after `MainloopInfo`

We need 3 or more stages for Ampere-style multistaged pipeline: 0th stage for `cp.async`, last stage for `mma.sync` and some `ld.shared`/`ldmatrix`, one before the last stage for load to first kgroup. 

We need 2 kgroups to interleave a kgroup of `ld.shared`/`ldmatrix` and different kgroup of `mma.sync`.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 4 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:551]: I see it function headers are with `///` and not `//`

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 3 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: sure. changed it to `getNvidiaAmpereTensorCorePipeline`

Added Ampere to the name as it is very specific to the Ampere architecture. Volta and Turing had 2-staged pipeline, Ampere started with 3-stages or more, and Hopper will have even different mainloop choreography. Thus, added Ampere to the name.

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp (hunk 2 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: So just remove this entire `if block` [iree/GPUUtils.cpp:L274](https://github.com/openxla/iree/blob/8ce3ac1aa8e1e8bfaa531a9b1cec011da6fb0d29/compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp#L281)

Is it not used somewhere else. I thought commenting it is a hack and we need a better fix for reducing the number of barrierOp after the prologue's cp.async.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: what I meant is remove the change. Do you need this? I assumed this is a hack you forgot to remove?

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 4 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp:551]: we follow this standard:
https://llvm.org/docs/CodingStandards.html#doxygen-use-in-documentation-comments

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 5 above)
Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: right, but can we just skip the pipelining instead of asserting (and doing the wrong thing if asserts are off)

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp (hunk 2 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: This will add a `bar.sync 0;` after every `cp.async` stage in the prologue. We don't need it for functionality, and it may add a performance overhead for matmuls with shorter mainloop a.k.a. smaller problem shape K. For performance, we really don't want to see `bar.sync 0;` in prologue for other than the last `cp.async` stage in the prologue. I have not quantified the performance penalty of keeping `bar.sync 0` `AroundSharedMemoryCopy` for every `cp.async` stage.

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: right but this is not correct unless we multi-buffer and pipeline in a certain way so we cannot just remove it here. That would cause potential functionality problems.

Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: If we keep this change, i.e., have `gpu::BarrierOp` Around every SharedMemoryCopy. 

```mlir
gpu.gpu_barrier;               // barrier at the start of Smem copy
nvgpu.cp.async;               // x4 for Tile A
nvgpu.cp.async;               // x4 for Tile B
gpu.gpu_barrier;              // barrier at the end of Smem copy
```

Then the pipeline and schedule for Ampere Tensor Cores `getNvidiaAmpereTensorCorePipeline` should ignore the one at the start of Smem copy. It is not needed for multi-staged pipeline. I think it is there to support single-stage pipelines, where the same stage is read and written?

Comment (ThomasRaoux) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: >  I think it is there to support single-stage pipelines, where the same stage is read and written?

Yes. We can remove the barrier when we pipeline but removing it here is not correct.

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 1 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: Thanks for the suggestion. 

This is possible under the assumption that `mma.sync`s are appearing in the mainloop in monotonically increasing kgroup order, i.e., for `numKgroup = 4`, we should see `mma.sync` for kgroup 0 followed by kgroup  1 and so on... In beginning of the endeavor, we collected more information and conservatively to get this schedule right. The assumption stated above holds and we can make the switch from `unordered_map` to `llvm::SmallVector`. I have pushed the change and verified the correctness.

File: compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp (hunk 2 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Utils/GPUUtils.cpp]: Removed the change. Thus, now we have two `gpu.gpu_barrier;`s when the IR hits scheduler/pipeliner, but we erase the `gpu. gpu_barrier ` that appears before the start of Shared Memory copy, as it is not required for the multi-staged Ampere-stype pipelines using `cp.async`

File: compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp (hunk 6 above)
Comment (manishucsd) [reply compiler/src/iree/compiler/Codegen/Common/GPUPipelining.cpp]: We have moved from conservatively collecting data to less data in a more precise data structures, provided the assumptions stated above hold true in our flow.

**Comment by manishucsd** [comment]
> Could you add a test case that demonstrates what the new code is doing? Maybe this already exists and in that case, you just forgot to update the check lines :).

Updated tests in compiler/src/iree/compiler/Codegen/LLVMGPU/test/nvvm_mma_sync_pipeline_test.mlir and compiler/src/iree/compiler/Codegen/Common/test/gpu_pipeline.mlir.

**Comment by manishucsd** [comment]
A variant of this has been merged. Closing this one.
